### ISSUE TITLE
Frequent AppVeyor Timeouts

### ISSUE THREAD

**Issue by Noordfrees** [widelands/widelands#3667]
The 64-bit debug builds on AppVeyor frequently fail due to a timeout. The other three builds take about one hour each, this one often timeouts after 2h.

**Comment by TinoM** [comment]
This is mainly due to:
- The memory/cpu limit on the build images
- The circular dependencies in widelands

We already have disabled parallel linking (otherwise it takes >8GB mem):
`-DCMAKE_JOB_POOLS=\"linking=1\" -DCMAKE_JOB_POOL_LINK=linking`

```
[01:09:31] [2006/2012] Generating wl_icon.o
[01:09:31] [2007/2012] Building CXX object src/CMakeFiles/widelands.dir/main.cc.obj
[01:22:46] [2008/2012] Linking CXX executable src\ai\test\test_ai.exe
[01:36:20] [2009/2012] Linking CXX executable src\widelands.exe
[01:49:26] [2010/2012] Linking CXX executable src\economy\test\test_economy.exe
```

So building takes the same time as x86 debug, but the linking takes ages...

**Comment by Noordfrees** [comment]
Perhaps we could skip building the tests on appveyor, since travis already runs them for us?

**Comment by gunchleoc** [comment]
We could skip the tests for the 64 bit debug builds only.

I have a feeling that our linking time became worse a while ago. Cleaning up the circular dependencies is very hard though, resolving them tends to snowball big time, so I already have a few dead-ended branches on the topic.

**Comment by hessenfarmer** [comment]
rather then skipping the tests I'd feel like we could speed up all our CI Builds by skipping the translations.   
If we compile them once in a seperate thread (like the code checking) ans push them back onto github they could be fetched by the builds on appveyor to be incorporated into the installer artefact. 
On travis we don't use them anyhow as we don't provide any installable binaries don't we.

**Comment by gunchleoc** [comment]
Travis builds won't time out due to translations, so we're fine there.
The reason we're building them on AppVeyor is to get them into the downloadable installers.

How about building them on AppVeyor for release builds only? This way, people can still play a current version in their language, and we save some time on the debug builds.

**Comment by hessenfarmer** [comment]
I know why we are doing this on Appveyor. And I really would like to keep the translations in the package as long as there is no other possibility.
That is why I thought about building them once and then just need to download them to incorporate them. 
Proposing this for travis as well was just to save time in the CI environment. As we (at least I) have to wait for a completed CI build with the last master merged into a branch to be able merge a branch, it is always a race who first get's his approved changes merged and who has to update the branch again. Saving time in the CI would speed this up a bit. 
So if it would be possible to have them build by perhaps another github action and pushed back to the branch we could download them on appveyor and save the time there and in travis to always build them.

**Comment by Noordfrees** [comment]
Pushing to master through actions is broken because the branch protection rules allow only admins to do that. Especially the translations update introduces one format issue every time. (I cannot look into this because all branch settings are admin-only.) 

What could be done though is to create another repo just for translation files which is updated by an action and pulled by appveyor. I could setup an i18n-repo and write the action easily if desired. Incorporating it into the builds needs to be done by someone who knows the appveyor scripting interface though.

---
> As we (at least I) 

(everyone except the admins)
> have to wait for a completed CI build with the last master merged into a branch to be able merge a branch, it is always a race who first get's his approved changes merged and who has to update the branch again.

yes, I also find this very annoying

**Comment by hessenfarmer** [comment]
Having a seperated repo is even better I believe. I would vote for it, but as GunChleoc is our expert in localization let's wait for her judgement.

**Comment by TinoM** [comment]
```
[00:08:36] cmake --build .
[00:08:37] [1/1985] Generating translations for locale de category scenario_tutorial01_basic_control.wmf
[...]
[00:10:21] [1405/1985] Generating translations for locale uk category scenario_fri02.wmf
[00:10:21] [1406/1985] Building CXX object src/CMakeFiles/build_info.dir/build_info.cc.obj
```
Doing the translations takes ~2 minutes, so skipping these won't help with the timeouts due to linking time.
Skipping tests would help (because linking the test executable binaries takes the same amount as linking widelands.exe) would defeat the whole purpose of the appveyor builds:
I still believe that windows x64 user are the biggest user group of widelands and we should take care that the code does build with mingw x64 and(!) runs fine.

**Comment by hessenfarmer** [comment]
ok so skipping translation is not a big benefit. Perhaps it might be worth a try to use the Hyper-V "4-cores" environment to save time on parallel building

**Comment by hessenfarmer** [comment]
another solution might be to try the bake image function to save time on downloading all MSYS updates every time

**Comment by gunchleoc** [comment]
I have a PR up to speed up linking a bit: https://github.com/widelands/widelands/pull/3720

It won't do much, bu it might just push us over the edge for some of the timeouts.

If we pre-build the translations, +1 for putting them in a separate repository.

> Especially the translations update introduces one format issue every time.

I noticed that too. I think my translation update script is missing a git add somewhere. Me & SirVer are the only people who can run this script, because you need to be an admin both here and at Transifex.

Does anybody want to volunteer to become admin on Transifex? You won't need to do anything, it's just that I don't like the bus factor.

ETA: The following PR will hopefully fix the formatting problems: https://github.com/widelands/widelands/pull/3722

**Comment by Noordfrees** [comment]
> Does anybody want to volunteer to become admin on Transifex? You won't need to do anything, it's just that I don't like the bus factor.

Okay, I volunteer :)  
(my username there is [Nordfriese](https://www.transifex.com/user/profile/Nordfriese/))

**Comment by hessenfarmer** [comment]
Tinos latest change gave us a big speedup on appveyor, which is still our CI Bottleneck. Thanks a lot for doing so. 
As my suggested function of baking an image is not cost free, we can't use it. However I didn't find out if caching some data would be helpful. If we could cache our MSYS/mingw folders we won't need so much time to download and install all the packages in there. This could give us another speedup in build time. From what I see caching is possible without any cost.

**Comment by TinoM** [comment]
Relating to the longest running x86_x64 build:
@gunchleoc overhaul of the includes saved us ~15 minutes build time, using the HyperV cloud another ~10 minutes (when i switched to GCE some time ago it was the other way round).

We are always updating the whole build environment (msys2) at the beginning because Appveyor does only update the images every few months. But downloading is done with internal mirrors, so i suspect we would only save ~2-3 minutes with the drawback that breaking dependency changes would be detected some weeks later.

Caching (so saving and reloading some produced data between builds) is a double-edged sword:
When do you invalidate the cache to avoid stale/obsolete data?

There are some discussions in the forums about using ccache on Appveyor and caching the cache of ccache (*g*) but i am not sure if the effort is worth the benefit...

**Comment by hessenfarmer** [comment]
My idea was to cache C:/msys64 each time. This would be extracted after getting the repository. As this would contain the db of MSYS2 our pacman commands would result in nothing to do or only small things to do that get cached after the build again. 
I agree that downloading does not take long however the current install takes up to 7 minutes from which possibly 5 could saved.

**Comment by gunchleoc** [comment]
I am manually cleaning up our external includes too, so that should have us additional memory & time.

One Pull Request is already up: https://github.com/widelands/widelands/pull/3724

Follow-up branches:
1. https://github.com/gunchleoc/widelands/tree/cleanup-includes2
2. https://github.com/gunchleoc/widelands/tree/cleanup-SDL-headers

**Comment by hessenfarmer** [comment]
For whatever reason we are timing out again. So we might try the caching solution probably. When I am back at my machine I could provide a branch to show you what I mean.

**Comment by hessenfarmer** [comment]
ok the reason is on msys2 side where repo.msys2.org is not reachable. But they are wotking on this.

**Comment by hessenfarmer** [comment]
made a branch to provide a temporary fix for current problems.
#3745

**Comment by hessenfarmer** [comment]
grrrrr... now sourceforge is failing however msys2-org seems to be online again.

**Comment by Noordfrees** [comment]
Will no longer be an issue after #4065
